File: c++/olimpiada/subiecte/oji2012/oji2012/Solutii9_12/Solutii9_12/OJI11_12/Subiecte_OJI_2012_XI_XII/blis/solutie/surse/PA_pblis.cpp
//sursa oficiala - liste implementate cu pointeri
//100 puncte
#include<cstdio>
#include<cstring>
using namespace std;
struct nod{int val,lg; nod *urm;};
nod *P[100010],*paux;
int k,n,i,v,j,K,ST,DR,MI,LMAX,sol,BST[100010],oo=2000000000;
char B[100010];
int main()
{
    freopen("blis.in","r",stdin);
    freopen("blis.out","w",stdout);
    scanf("%d",&k);
    scanf("%s",B+1);
    n=strlen(B+1);
    for(i=1;i<=n+1;i++)BST[i]=oo;BST[0]=-oo;LMAX=0;
    for(i=1;i<=n;i++)
    {
        v=0;K=i+k-1<n?i+k-1:n;
        for(j=i;j<=K;j++)
        {
            v<<=1;v|=B[j]-'0';sol=v>sol?v:sol;
            for(ST=0,DR=LMAX+1;DR-ST-1;)
            {
                MI=(ST+DR)>>1;
                if(BST[MI]<v)ST=MI;
                else DR=MI;
            }
            if(v<BST[DR])
            {
                paux=new nod;paux->val=v;paux->lg=DR;paux->urm=P[j];P[j]=paux;
            }
        }
        for(;P[i];)
        {
            paux=P[i];P[i]=P[i]->urm;
            if(BST[paux->lg]>paux->val)BST[paux->lg]=paux->val;
            delete paux;
        }
        while(BST[LMAX+1]<oo)LMAX++;
    }
    printf("%d\n%d",sol,LMAX);
    return 0;
}

File: c++/Contests/olimpiada/subiecte/oji2012/oji2012/Solutii9_12/Solutii9_12/OJI11_12/Subiecte_OJI_2012_XI_XII/blis/solutie/surse/PA_pblis.pas
//sursa oficiala - liste implementate cu pointeri
//100 puncte

type pnod = ^nod;
	  nod = record
				val,lg:longint;
				urm:pnod;
			end;

const oo=2000000000;

var paux:pnod;
	k,n,i,v,j,km,st,dr,mi,lmax,sol:longint;
	p:array[0..100010] of pnod;
	bst:array[0..100010] of longint;
	b:array[0..100010] of char;
	fin,fout:text;

begin
    assign(fin,'blis.in');
	reset(fin);
	
	readln(fin,k);

	n:=0;
	while (not eof(fin)) do
		begin
			n:=n+1;
			read(fin,b[n]);
		end;
		
	close(fin);
  
    for i:=1 to n+1 do bst[i]:=oo;
	bst[0]:=-oo;
	lmax:=0;
	sol:=0;
    
	for i:=1 to n do
		begin
			v:=0;
			if i+k-1<n then km:=i+k-1
			else km:=n;
        
			for j:=i to km do
				begin
					v:=v shl 1;
					v:= v or (ord(b[j])-ord('0'));
					if v>sol then sol:=v;
					
					st:=0;
					dr:=lmax+1;
					while dr-st-1<>0 do
						begin
							mi:=(st+dr) shr 1;
							if bst[mi]<v then st:=mi
							else dr:=mi;
						end;
					if v<bst[dr] then
						begin
							new(paux);
							paux^.val:=v;
							paux^.lg:=dr;
							paux^.urm:=p[j];
							p[j]:=paux;
						end;
				end;
        while p[i]<>nil do
			begin
				paux:=p[i];
				p[i]:=p[i]^.urm;
				if bst[paux^.lg]>paux^.val then bst[paux^.lg]:=paux^.val;
				dispose(paux);
			end;
        
		while bst[lmax+1]<oo do lmax:=lmax+1;
    end;
	
	assign(fout,'blis.out');
	rewrite(fout);
	
	writeln(fout,sol);
	write(fout,lmax);
	
	close(fout);

end.
